Allow http.HTTPStatus enums as response status codes.

jkrukoff opened the pull request:
Python 3.5 introduced a new enumeration "http.HTTPStatus" for
representing HTTP response status codes. The default response validation
introduced in connexion 1.1.12 highlighted the fact that connexion does
not natively support this type and was previously silently ignoring
non-integer status code representations.

This modifies the response validation code to extract the value when
given an enum instead of an int. Somewhat hacky test code is added to
check for enum support on python versions that include
"http.HTTPStatus".

Fixes #503 



Changes proposed in this pull request:

 - Convert enums to ints in response validation matching.

jmcs commented:
I like it, but can you had a unit test for custom enum (this would cover the case where someone uses a backport of the enum).

jkrukoff commented:
Do you have a suggestion on a good way to only run the test in python versions where Enum is available, or should I use the same ImportError based approach?

mvalkon commented:
@jkrukoff there's a way to do this yes. First you need to indicate that the test should be skipped in python versions that do not support enums. You can use the [@unittest.skipIf][1] decorator:

```python
@unittest.skipIf(sys.version_info < (3, 5), "Not supported in this version")
def test_your_test_name(...):
```

If you want to run your tests with multiple python versions, we're using [tox][2] for this. To verify that the above works you can run tox with `-e` command line parameter as follows:

```bash
$ tox -e py27,py34,py36  
```

You can install tox with `pip`:

```bash
$ pip install tox
```
tox is currently not in https://github.com/zalando/connexion/blob/master/requirements-devel.txt but I'll make a PR to add it.

[1]: https://docs.python.org/3/library/unittest.html#unittest.skipIf
[2]: https://tox.readthedocs.io/en/latest/

jkrukoff commented:
Added the test for arbitrary enums, which exposed an assumption inside flask that appears to only work for ints/actual HTTPStatus Enums. Moved the enum value extraction deeper into the stack to compensate.

Tried to use @unittest.skipif in the test suite, but that failed with an unsupported error from unittest for python versions <3.5. Reverted back to the previous approach to get tests passing again.

mvalkon commented:
@jkrukoff `unittest.skipIf` has been introduced in 3.1. Out of interest, what sort of an exception did you get?

The change looks good to me.

jkrukoff commented:
It was pretty close to "unittest.skipif is unsupported in this version". It it's been present since 3.1, would explain why the python 2.7 test suite was exploding with that error.

jmcs commented on tests/fakeapi/hello.py:
```diff
@@ -1,4 +1,6 @@
 #!/usr/bin/env python3
+import sys
+import unittest
```
This imports are not being used.

jkrukoff replied:
Ah, yeah, forgot to remove those from the attempt to use uniftest.skipIf.

jmcs commented on connexion/apis/flask_api.py:
```diff
@@ -154,6 +154,12 @@ def _build_flask_response(cls, mimetype=None, content_type=None,
         flask_response = flask.current_app.response_class(**kwargs)  # type: flask.Response
 
         if status_code is not None:
+            try:
```
Please use ``hasattr`` here since the most common use case is ``status_code`` being an integer.

[``hasattr`` is faster than ``try... except AtributeError`` for cases where the attribute doesn't exist](https://stackoverflow.com/a/903238).

jkrukoff replied:
Yes, and you get to repeat yourself while doing the check. It'd be even faster if just didn't support enums at all!

Will update.

hjacobs commented:
:+1:
